mesh.py

```diff
@@ -34,17 +43,189 @@ def add_face(self, indices):
 
-    def write_obj(self, file):
+    def write_obj(self, obj_file):
         for (x, y, z) in self.vertices:
-            file.write(f'v {x} {y} {z}\n')
+            obj_file.write(f'v {x} {y} {z}\n'.encode('utf-8'))
 
         for (nx, ny, nz) in self.normals:
-            file.write(f'vn {nx} {ny} {nz}\n')
+            obj_file.write(f'vn {nx} {ny} {nz}\n'.encode('utf-8'))
 
         for [v1, v2, v3, n] in self.faces:
             v1_idx = v1 + 1
             v2_idx = v2 + 1
             v3_idx = v3 + 1
             n_idx = n + 1
-            file.write(
-                f'f {v1_idx}//{n_idx} {v2_idx}//{n_idx} {v3_idx}//{n_idx}\n')
+            line = f'f {v1_idx}//{n_idx} {v2_idx}//{n_idx} {v3_idx}//{n_idx}\n' 
+            obj_file.write(line.encode('utf-8'))
+
+    def make_glb_buffer_views(self):
+        vertex_bv = b''
+        normal_bv = b''
+
+        min_x = float('inf')
+        min_y = float('inf')
+        min_z = float('inf')
+        max_x = -float('inf')
+        max_y = -float('inf')
+        max_z = -float('inf')
+
+        for [v1, v2, v3, n] in self.faces:
+            (x, y, z) = self.vertices[v1]
+            vertex_bv += pack_vec3(x, y, z)
+            min_x = min(min_x, x)
+            min_y = min(min_y, y)
+            min_z = min(min_z, z)
+            max_x = max(max_x, x)
+            max_y = max(max_y, y)
+            max_z = max(max_z, z)
+
+            (x, y, z) = self.vertices[v2]
+            vertex_bv += pack_vec3(x, y, z)
+            min_x = min(min_x, x)
+            min_y = min(min_y, y)
+            min_z = min(min_z, z)
+            max_x = max(max_x, x)
+            max_y = max(max_y, y)
+            max_z = max(max_z, z)
+
+            (x, y, z) = self.vertices[v3]
+            vertex_bv += pack_vec3(x, y, z)
+            min_x = min(min_x, x)
+            min_y = min(min_y, y)
+            min_z = min(min_z, z)
+            max_x = max(max_x, x)
+            max_y = max(max_y, y)
+            max_z = max(max_z, z)
+
+            (nx, ny, nz) = self.normals[n]
+            normal_bin = pack_vec3(nx, ny, nz)
+            normal_bv += normal_bin * 3
+
+        min_pos = [min_x, min_y, min_z]
+        max_pos = [max_x, max_y, max_z]
+        return vertex_bv, normal_bv, min_pos, max_pos
+
+    def write_glb(self, glb_file):
+        GLB_ALIGNMENT = 4
+        GLB_MAGIC = b'glTF'
+        GLB_LEN_HEADER = 12
+        GLB_CHUNK_BIN = b'BIN\x00'
+        GLB_CHUNK_JSON = b'JSON'
+        GLB_VERSION = pack_u32(2)
+        GLB_FLOAT = 5126
+        
+        vertex_bv, normal_bv, min_pos, max_pos = self.make_glb_buffer_views()
+        
+        vert_offset = 0
+        vert_len = len(vertex_bv)
+        vertex_bv = pad_binary(vertex_bv, GLB_ALIGNMENT)
+        
+        normal_offset = len(vertex_bv)
+        normal_len = len(normal_bv)
+        normal_bv = pad_binary(normal_bv, GLB_ALIGNMENT)
+
+        buf = vertex_bv + normal_bv
+        
+        gltf_json = {
```

Comment:
Also try loading the result in something other than Blender. Perhaps Cesium? 😉 See https://sandcastle.cesium.com/?src=3D%20Models.html

Reply:
Tried it with [this sandcastle](https://sandcastle.cesium.com/#c=tVZtb9s2EP4rnD8pgCbZeWkz1wmW2Fs7bEG72i0GzEPASKeICEUKJOUkK/Lfe3xRLL+0mBfYX0zeHZ97niMp3oIqsmBwD4qcEQH3ZAyaNVXy2dmieS9z87EUhjIBat6LyZe5IISJQl7KhyEpKNcQW5MGDplhUvwmcpZRI9WKt6S5vNdDYlTTWmTD8wvBKmrg2f508GYu5qJohMMimQJ0X8kceNQoHpMS2G1pDjwNzz0BYZhhoBMFlVzABeeRh8EIVFhLzRzYWatvTJXBERVHSaFkNYFbBaAju4CQHweHR0n/9fHxq8FPsTcdHyf9k/7R6/6rYPAs7Ngm8mlKoDkTt8ssV9SUiZEf0UyFjgZHJ8vgmpmsxND+s0VJzlcMZb22K+98gg926UeMjkLG2KPFDmKZQiqGhaGrymcKqRRSVTop1+D+bLDSSmB4KERbt1ZzrVrBbQpX+EdEX98ImufRF79M0Aq31+7dKupwHb/Dd9idBHdlz8CQBFSCiKwLiwEMz1JTfWAPwKfsX0w6ODxdeumD9U4zytFz2Mdf8D25/6dQOC/EKJrdQf5Lq88LxYgnK95Vt7bcNPr+tusCLQMPZkjmvSlQXQLneGGCOOHvB96JcLSjg6WW7ilvbQRhSmPqYZpyiaxLqc3wtH/aTysqSmoMFdeZFJDc8pvnNPY36LdjLwm1OYHxFp4XTGWKFuY7PMl/IZok6ZRWNYcJNTR1W6VTf+QwRRhd43CD7QluRLLJOP4m5YlC0Xvk6/C7480Cn+zG+K2SjcjJZyhZxvdD3acIGVZnG/R3of5OGoK7Ri4p51KKPZY9ZFidbTnbO56WK8bvyEw12d0eudskLsf6/EW1n94xISAn45Li18i413dvCmhb+Wscvoj2BMlKMpZVrUBr5O+I7IW7S7XMtKX6pvjfOn6f/XXYlbHjFWAFiX4Ij+6vqKlRMAHjO6REN3UtldGXVDMdhTdHZyDgoANByD0T2DUl+GQp0ymHLcisZJrcKOypsHnLJWgi8K4GYOKAySfBFqA05SRb6rD6kIye95aAbTnCU/GyK2wLt2nZ+Sr/4/qMKRV5RrXB7xj2FDMp+Q1VVyCaKDzAdmEv7o20eeRw3sL9zCpXB+wPImRsABmjHJ3e4LkAk2RatxlHaXfpKGcLwvKzLa0vyTjVGj1Fw12DMe+dj1KM31jKpeus3mPxOX20YeXg/A9vTJJklOJ0+0rjBa4hfwU) (Requires a static server). It works, though this makes it obvious that I'm not handling the coordinate system correctly.

![image](https://user-images.githubusercontent.com/8422414/125776921-440adcfc-9429-48cb-82d6-ddbdf5741904.png)


The [glTF spec](https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#coordinate-system-and-units) says:
>glTF uses a right-handed coordinate system, that is, the cross product of +X and +Y yields +Z. glTF defines +Y as up. The front of a glTF asset faces +Z.

I've been using a z-up coordinate system the whole time 😅 

I should create a `logbook.md` like I do with other repos and document the things I should do in the future:
1. fix coordinate system
2. Apply materials and texture coordinates
3. Other future work